### PR TITLE
Prohibit pytest.mark.usefixtures in pytest.param

### PR DESCRIPTION
**Pull request by Hawk777**

### CODE REVIEW

**Comment by Hawk777** [comment]
I’m not certain this is the best place for this change. As far as I know, `pytest.param` is only useful to either pass to `pytest.mark.parametrize` or to `Metafunc.parametrize`, and `usefixtures` doesn’t work in either of those places, so this seemed logical to me, but I don’t know Pytest’s architecture.

**Comment by Hawk777** [comment]
See GH-4112.

**Review by The-Compiler — COMMENTED** [review]
Seems reasonable to me, and definitely is a step forward from silently ignoring those. Could you add a test case for this as well? `TestMarkersWithParametrization` in `testing/python/metafunc.py` seems like the right place to me.

**Comment by Hawk777** [comment]
I can’t seem to run the tests locally (`ERROR: /tmp/pytest/pyproject.toml: 'minversion' requires pytest-2.0, actual pytest-0.1.dev1+gc40a0d5'`), so I’ve tried to do that in 0cb2599d5b35cc9d6c4ded65ae1959039e5f30b6 and I’ll let CI see if I got it right.

**Comment by Hawk777** [comment]
Looks like a pass. That fixup commit should probably be squashed once you’ve reviewed it (the merge/squash guidelines in `CONTRIBUTING.md` say that you’ll do that, but let me know if you prefer that I do).

**Comment by Hawk777** [comment]
@The-Compiler just checking if there’s anything else you were waiting for from me? Not sure if you get notifications or not for messages that don’t ping you. No hurry obviously.

**Comment by The-Compiler** [comment]
Sorry for the radio silence! Seems fine to me, though I'd like a second pair of eyes on the question whether this is a good place to do this kind of thing. @RonnyPfannschmidt @nicoddemus any opinions?

File: testing/python/metafunc.py
Comment (nicoddemus) [inline testing/python/metafunc.py:2123]: I think it would be better for us to have an integration test instead, because in this it is important for us to see how the error will appear to the user: for example, it should not blow up an entire traceback with the pytest internals, it should show a succinct message pointing out the fixture, location, etc.

I would start by using `pytest.fail(..., pytrace=False)` or raising `UsageError`, whichever produces a better result.

File: changelog/4112.improvement.rst
Comment (nicoddemus) [inline changelog/4112.improvement.rst:1]: I suggest we tweak this a bit:

```suggestion
Using :ref:`pytest.mark.usefixtures <pytest.mark.usefixtures ref>` on :func:`pytest.param` now produces an error instead of silently doing nothing.
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @Hawk777 this is a great start!

Please take a look at my comments.

File: testing/python/metafunc.py
Comment (Hawk777) [reply testing/python/metafunc.py:2123]: @nicoddemus `UsageError` and `ValueError` look the same to me, like this (other than the exception name, obviously):

```
___________________ ERROR collecting test_foo.py _____________________
test_foo.py:6: in <module>
    pytest.param(
/usr/lib/python3.12/site-packages/_pytest/mark/__init__.py:72: in param
    return ParameterSet.param(*values, marks=marks, id=id)
/usr/lib/python3.12/site-packages/_pytest/mark/structures.py:92: in param
    raise ValueError(
E   ValueError: pytest.param cannot add pytest.mark.usefixtures; see https://github.com/pytest-dev/pytest/issues/4112
```

while `pytest.fail` looks like this:

```
___________________ ERROR collecting test_foo.py _____________________
pytest.param cannot add pytest.mark.usefixtures; see https://github.com/pytest-dev/pytest/issues/4112
```

So it looks like `pytest.fail` produces the least output (and without any Pytest internals mentioned). However, it also doesn’t mention the line number on the user test file where the error occurred (remember this happens at import time, not execution time, so it also doesn’t mention the function name), which the others (since they include traces) do. Is that a problem? It seems to me it would be useful to point the user at the specific `param` call that has the problem, but if you prefer the shorter output, I’ll go ahead with that.

So, given the above tradeoff, what’s your preference? `ValueError`, `UsageError`, or `pytest.fail`?

Comment (nicoddemus) [reply testing/python/metafunc.py:2123]: Hmm I see thanks.

Let's keep it `ValueError` for now then, I agree that the location/line number is really important -- due to how `pytest.param` is exposed in the API, this is reasonable to do for now (we could do some `sys.frame()` shenaningans to get the `pytest.param` location invocation and build up a message ourselves, but not sure it is worth doing that for this quite uncommon occurrence).

But we should move forward with the integration test I mentioned: it is important that we check that `test_foo.py:6` and `pytest.param(` appear in the message (from your example posted above).

Also, we probably should put a link to the docs somewhere explaining that (possibly we can add a note to `pytest.param` documentation), rather than pointing to the GitHub issue.

Comment (Hawk777) [reply testing/python/metafunc.py:2123]: Yep, I will definitely do the integration test change, just waiting to do that until we’ve nailed down what we want the behaviour to be (no point changing the test twice).

In terms of a link to documentation, my understanding was that one *should* be able to `usefixtures` in `param`, but it just didn’t work right now due to a bug. Is it actually intended that `usefixtures` not be used in `param`? Or do you think, even though it’s a bug, it should be in the manual (typically I don’t see bugs documented in manuals, but rather only bug trackers, on most projects)?

Comment (nicoddemus) [reply testing/python/metafunc.py:2123]: > Yep, I will definitely do the integration test change, just waiting to do that until we’ve nailed down what we want the behaviour to be (no point changing the test twice).

Oh sure, thanks, just wanted to take the opportunity to point out what I would like to see checked in the integration test.

> Is it actually intended that usefixtures not be used in param?

We don't have any plans to support `usefixtures` there (nor in fixtures), so we might as well place the error (as you are doing) and leave a note in the docs (even if that changes in the future, I think it is harmless to let users know the current situation explicitly).

File: src/_pytest/mark/structures.py
Comment (nicoddemus) [inline src/_pytest/mark/structures.py:94]:
```suggestion
                "https://docs.pytest.org/en/stable/reference/reference.html#pytest-param"
```

**Review by nicoddemus — APPROVED** [review]
Thanks! 🙇

File: testing/python/metafunc.py
Comment (Hawk777) [reply testing/python/metafunc.py]: I was hoping to make fixing this one commit, but I had no idea CI was going to come in and add extra commits, so that didn’t work out. So this is now b3f232a800108c32368c030dcd7b39990240188a and ccc1e18702117d5671f5553a420eadef68405edb, plus anything else that needs to follow it.

**Comment by Hawk777** [comment]
The warnings in the docs build AFAICT are unrelated to this patch, so I think this is good to go now?

**Comment by nicoddemus** [comment]
It is related:

```
 /home/docs/checkouts/readthedocs.org/user_builds/pytest/envs/12828/lib/python3.12/site-packages/_pytest/mark/__init__.py:docstring of _pytest.mark.param:21: WARNING: undefined label: 'pytest.mark.usefixtures' [ref.ref]
```

But I just fixed it!

I will leave it open for a few more days to give others a chance to review it, otherwise I will merge it.

Thanks again!

**Comment by Hawk777** [comment]
Oh, sorry, somehow I didn’t see that message. I only saw things about Spinx 90 compatibility issues in CSS and Javascript references, or something like that (I can’t find the page any more now that the new commit is running in CI). Thanks for the review!

**Pull request merged**
